[DX-1525 TT-12662] endpoint rate limit (#5110)
TykTechnologies/tyk-docs: tyk-docs/content/product-stack/tyk-gateway/middleware/endpoint-rate-limit-oas.md

---

> title: Using the Rate Limit Middleware with Tyk OAS APIs

Reviewer: ```suggestion
title: Using the Rate Limit middleware with Tyk OAS APIs
```

date: 2024-07-18
description: "Using the per-endpoint rate limit middleware with Tyk OAS APIs"
tags: ["rate limit", "middleware", "per-endpoint", "Tyk OAS", "Tyk OAS API"]
---

The per-endpoint rate limit middleware allows you to enforce rate limits

> on specific endpoints. This middleware is configured in the Tyk OAS API

Reviewer: ```suggestion
on specific endpoints. This middleware is configured in the [Tyk OAS API
```

> Definition, either via the Tyk Dashboard API or in the API Designer.

Reviewer: ```suggestion
Definition](https://tyk.io/docs/tyk-apis/tyk-gateway-api/oas/x-tyk-oas-doc/#operation), either via the Tyk Dashboard API or in the API Designer.
```

> ## Configuring a Rate Limit in the Tyk OAS API Definition

Reviewer: ```suggestion
## Configuring a rate limit in the Tyk OAS API Definition
```

The design of the Tyk OAS API Definition takes advantage of the
`operationId` defined in the OpenAPI Document that declares both the path
and method for which the middleware should be added. Endpoint `paths`
entries (and the associated `operationId`) can contain wildcards in the
form of any string bracketed by curly braces, for example
`/status/{code}`. These wildcards are so they are human-readable and do
not translate to variable names. Under the hood, a wildcard translates to
the “match everything” regex of: `(.*)`.

> The rate limit middleware can be added to the `operations` section of the

Reviewer: ```suggestion
The rate limit middleware (`rateLimit`) can be added to the `operations` section of the
```

Tyk OAS Extension (`x-tyk-api-gateway`) in your Tyk OAS API Definition
for the appropriate `operationId` (as configured in the `paths` section
of your OpenAPI Document).

> The `rate_limit` object has the following configuration:

Reviewer: ```suggestion
The `rateLimit` object has the following configuration:
```

- `enabled`: enable the middleware for the endpoint

> - `rate`: the number of requests allowed

Reviewer: ```suggestion
- `rate`: the maximum number of requests that will be permitted during the interval (window)
```

> - `per`: the time period in seconds

Reviewer: ```suggestion
- `per`: the length of the interval (window) in time duration notation (e.g. 10s)
```

#### Simple endpoint rate limit

For example:

> ~~~

Author: @andyo-tyk just to clarify, you want to intentionally highlight lines 39-43, should a similar highlight block be defined for classic?

{
    "components": {},
    "info": {
        "title": "example-rate-limit",
        "version": "1.0.0"
    },
    "openapi": "3.0.3",
    "paths": {
        "/status/200": {
            "get": {
                "operationId": "status/200get",
                "responses": {
                    "200": {
                        "description": ""
                    }
                }
            }
        }
    },
    "x-tyk-api-gateway": {
        "info": {
            "name": "example-rate-limit",
            "state": {
                "active": true
            }
        },
        "upstream": {
            "url": "http://httpbin.org/"
        },
        "server": {
            "listenPath": {
                "value": "/example-rate-limit/",
                "strip": true
            }
        },
        "middleware": {
            "operations": {
                "status/200get": {

>                     "rate_limit": {

Reviewer: ```suggestion
                    "rateLimit": {
```

                        "enabled": true,
                        "rate": 60,

>                         "per": 1

Reviewer: ```suggestion
                        "per": "1s"
```

                    }
                }
            }
        }
    }
}

> ~~~

Reviewer: ```suggestion
```
```

Reviewer: Again, this should be ```

In this example Tyk OAS API definition, the rate limit has been
configured for the `GET /status/200` endpoint, limiting requests to 60
per second.

> 

Reviewer: ```suggestion

The configuration above is a complete and valid Tyk OAS API Definition that you can import into Tyk to try out the Per-endpoint Rate Limiter middleware.
```

#### Advanced endpoint rate limit

For more complex scenarios, you can configure rate limits for multiple

> paths. The order of evaluation matches the order defined in the

Reviewer: ```suggestion
paths. The order of evaluation matches the order that endpoints are defined in the
```

> `rate_limit` array. For example, to limit POST requests to `/user/login`

Reviewer: ```suggestion
`paths` section of the OpenAPI description. For example, if you wanted to limit the rate of `POST` requests to your API allowing a higher rate to one specific endpoint you could configure the API definition as follows: 
```

to 100 requests per second and apply a regex pattern for other endpoints:

~~~
{
    "components": {},
    "info": {
        "title": "advanced-rate-limit",
        "version": "1.0.0"
    },
    "openapi": "3.0.3",
    "paths": {
        "/user/login": {
            "post": {
                "operationId": "user/loginget",
                "responses": {
                    "200": {
                        "description": ""
                    }
                }
            }
        },
        "/{any}": {

>             "get": {

Reviewer: ```suggestion
            "post": {
```

>                 "operationId": "anyget",

Reviewer: ```suggestion
                "operationId": "anypost",
```

                "responses": {
                    "200": {
                        "description": ""
                    }
                }
            }
        }
    },
    "x-tyk-api-gateway": {
        "info": {
            "name": "advanced-rate-limit",
            "state": {
                "active": true
            }
        },
        "upstream": {
            "url": "http://httpbin.org/"
        },
        "server": {
            "listenPath": {
                "value": "/advanced-rate-limit/",
                "strip": true
            }
        },
        "middleware": {
            "operations": {
                "user/loginget": {
                    "rate_limit": {
                        "enabled": true,
                        "rate": 100,

>                         "per": 1

Reviewer: ```suggestion
                        "per": "1s"
```

                    }
                },

>                 "anyget": {

Reviewer: ```suggestion
                "anypost": {
```

                    "rate_limit": {
                        "enabled": true,
                        "rate": 60,
                        "per": 1
                    }
                }
            }
        }
    }
}
~~~

> In this advanced example, the first rule limits POST requests to

Reviewer: ```suggestion
In this example, the first rule limits requests to the
```

`/user/login` to 100 requests per second. Any other GET request matching

> the regex pattern `/{any}` will be limited to 60 requests per second. The

Author: rejecting this, should be RPS but i don't feel it adds something by abbreviating

> order of evaluation ensures that the specific `/user/login` endpoint is

Reviewer: ```suggestion
order of evaluation ensures that the specific `POST /user/login` endpoint is
```

> matched and evaluated before the regex pattern.

Reviewer: ```suggestion
matched and evaluated before the regex pattern.

The configuration above is a complete and valid Tyk OAS API Definition that you can import into Tyk to try out the Per-endpoint Rate Limiter middleware.
```

---
title: Using the Rate Limit middleware with Tyk OAS APIs
date: 2024-07-18
description: "Using the per-endpoint rate limit middleware with Tyk OAS APIs"
tags: ["rate limit", "middleware", "per-endpoint", "Tyk OAS", "Tyk OAS API"]
---

The per-endpoint rate limit middleware allows you to enforce rate limits
on specific endpoints. This middleware is configured in the [Tyk OAS API
Definition](https://tyk.io/docs/tyk-apis/tyk-gateway-api/oas/x-tyk-oas-doc/#operation), either via the Tyk Dashboard API or in the API Designer.

> 

Reviewer: ```suggestion

IIf you’re using the legacy Tyk Classic APIs, then check out the [Tyk Classic](https://tyk.io/docs/product-stack/tyk-gateway/middleware/endpoint-rate-limit-classic/) page.

```

        },
        "server": {
            "listenPath": {
                "value": "/example-rate-limit/",
                "strip": true
            }
        },
        "middleware": {
            "operations": {
                "status/200get": {
                    "rateLimit": {
                        "enabled": true,
                        "rate": 60,
                        "per": "1s"
                    }
                }
            }
        }
    }
}
```

In this example, a rate limit has been configured for the `GET
/status/200` endpoint, limiting requests to 60 per second.

The configuration above is a complete and valid Tyk OAS API Definition
that you can import into Tyk to try out the Per-endpoint Rate Limiter
middleware.

#### Advanced endpoint rate limit

For more complex scenarios, you can configure rate limits for multiple
paths. The order of evaluation matches the order that endpoints are
defined in the `paths` section of the OpenAPI description. For example,
if you wanted to limit the rate of `POST` requests to your API allowing a
higher rate to one specific endpoint you could configure the API
definition as follows: 

> ```json {hl_lines=["47-51", "54-58"],linenos=true, linenostart=1}

Reviewer: ```suggestion
```json {hl_lines=["49-53", "56-60"],linenos=true, linenostart=1}
```

{
    "components": {},
    "info": {
        "title": "advanced-rate-limit",
        "version": "1.0.0"
    },
    "openapi": "3.0.3",
    "paths": {
        "/user/login": {
            "post": {

>                 "operationId": "user/loginget",

Reviewer: ```suggestion
                "operationId": "user/loginpost",
```

Sorry, think I missed this before - the operationId usually merges path with method

                "responses": {
                    "200": {
                        "description": ""
                    }
                }
            }
        },
        "/{any}": {
            "post": {
                "operationId": "anypost",
                "responses": {
                    "200": {
                        "description": ""
                    }
                }
            }
        }
    },
    "x-tyk-api-gateway": {
        "info": {
            "name": "advanced-rate-limit",
            "state": {
                "active": true
            }
        },
        "upstream": {
            "url": "http://httpbin.org/"
        },
        "server": {
            "listenPath": {
                "value": "/advanced-rate-limit/",
                "strip": true
            }
        },
        "middleware": {
            "operations": {

>                 "user/loginget": {

Reviewer: ```suggestion
                "user/loginpost": {
```

To match the correction above

                }
            }
        }
    },
    "x-tyk-api-gateway": {
        "info": {
            "name": "advanced-rate-limit",
            "state": {
                "active": true
            }
        },
        "upstream": {
            "url": "http://httpbin.org/"
        },
        "server": {
            "listenPath": {
                "value": "/advanced-rate-limit/",
                "strip": true
            }
        },
        "middleware": {
            "operations": {
                "user/loginpost": {
                    "rateLimit": {
                        "enabled": true,
                        "rate": 100,
                        "per": "1s"
                    }
                },
                "anypost": {
                    "rateLimit": {
                        "enabled": true,
                        "rate": 60,
                        "per": "1s"
                    }
                }
            }
        }
    }
}
```

In this example, the first rule limits requests to the `POST /user/login`
endpoint to 100 requests per second (rps). Any other `POST` request to an
endpoint path that matches the regex pattern `/{any}` will be limited to

> 60 rps. The order of evaluation ensures that the specific `POST

Reviewer 2: Is this order of evaluation taken care of by the library?
